File: backend/app/schemas/threat.py
"""ICT Register Threat and Threat<->Risk Link relation schemas (issue #47).

Write schemas carry the workbook's entered 12_Hrozby columns only and forbid
unknown keys. The category is validated against the workbook closed list
``KategorieHrozeb`` from ``app.services._ict_register_reference``. Threats
carry no derived block — they sit outside the criticality cascade — so the
Read schema is the entered columns plus the archive state and per-row
capabilities (ADR-001).

The Threat<->Risk Link relation is manageable from BOTH ends: the Threat page
mutates under ``threats:write`` and the Risk detail mutates under
``risks:write`` (the #46 managing-end precedent applied to each end), so the
link row's ``can_delete`` capability is computed per serialization surface.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, computed_field, field_validator

from app.core.datetime_utils import UtcAwareDatetime
from app.schemas.collection import CollectionGroupRead
from app.services._ict_register_reference import THREAT_CATEGORY_CODES


class ThreatWriteValidators(BaseModel):
    """Shared closed-list enforcement for Threat write payloads."""

    model_config = {"extra": "forbid"}

    @field_validator("category", check_fields=False)
    @classmethod
    def _validate_category(cls, value: str | None) -> str | None:
        if value is not None and value not in THREAT_CATEGORY_CODES:
            raise ValueError("Value must be a canonical Threat category code")
        return value


class ThreatBase(ThreatWriteValidators):
    name: str = Field(..., min_length=1, max_length=255)
    threat_steward_user_id: int = Field(..., ge=1)
    category: str | None = Field(None, max_length=50)
    description: str | None = None
    typical_weaknesses: str | None = None
    relevant_subject: str | None = Field(None, max_length=255)
    notes: str | None = None


class ThreatCreate(ThreatBase):
    pass


class ThreatUpdate(ThreatWriteValidators):
    name: str | None = Field(None, min_length=1, max_length=255)
    threat_steward_user_id: int | None = Field(None, ge=1)
    category: str | None = Field(None, max_length=50)
    description: str | None = None
    typical_weaknesses: str | None = None
    relevant_subject: str | None = Field(None, max_length=255)
    notes: str | None = None


class ThreatCapabilities(BaseModel):
    can_read: bool
    can_update: bool
    can_archive: bool
    can_restore: bool


class ThreatStewardRead(BaseModel):
    """Safe display projection; intentionally excludes the database user id."""

    name: str
    email: str
    role_name: str
    department_name: str | None = None


class ThreatRead(BaseModel):
    id: int
    threat_steward_user_id: int | None = None
    threat_steward: ThreatStewardRead | None = None
    steward_orphaned: bool = False
    stewardship_status: Literal[
        "assigned",
        "legacy_unassigned",
        "pending_governance",
        "invalid_assignment",
    ] = "assigned"

    name: str
    category: str | None = None
    description: str | None = None
    typical_weaknesses: str | None = None
    relevant_subject: str | None = None
    notes: str | None = None
    is_archived: bool = False
    archived_at: UtcAwareDatetime | None = None
    archived_by_id: int | None = None
    capabilities: ThreatCapabilities | None = None
    created_at: UtcAwareDatetime
    updated_at: UtcAwareDatetime

    model_config = {"from_attributes": True}


class ThreatListCapabilities(BaseModel):
    """Collection-level Threat list action capabilities."""

    can_create: bool
    can_export: bool = False


class ThreatFacetOption(BaseModel):
    """One permission-scoped Threat facet option."""

    value: str
    label: str
    count: int
    disabled: bool = False
    selected: bool = False


class ThreatLookupOption(BaseModel):
    """Safe remote Threat-filter lookup; hidden Risk context is excluded."""

    id: int
    label: str
    secondary_label: str | None = None
    disabled: bool = False
    count: int | None = None


class ThreatListItem(ThreatRead):
    """Threat list projection with permission-scoped linked-Risk metadata."""

    # Hidden Risk links never contribute to this projection.
    visible_linked_risk_count: int = 0


class ThreatListResponse(BaseModel):
    items: list[ThreatListItem]
    total: int
    offset: int
    limit: int
    capabilities: ThreatListCapabilities | None = None
    groups: list[CollectionGroupRead] = Field(default_factory=list)
    facets: dict[str, list[ThreatFacetOption]] = Field(default_factory=dict)

    @computed_field
    def skip(self) -> int:
        return self.offset


class ThreatRiskLinkCreate(BaseModel):
    """Threat-end create payload: link this Threat to a Risk."""

    model_config = {"extra": "forbid"}

    risk_id: int = Field(..., ge=1)


class RiskThreatLinkCreate(BaseModel):
    """Risk-end create payload: link this Risk to a Threat."""

    model_config = {"extra": "forbid"}

    threat_id: int = Field(..., ge=1)


class ThreatRiskLinkCapabilities(BaseModel):
    """Per-row link actions: mutations follow the MANAGING end's write permission."""

    can_delete: bool


class ThreatRiskLinkRead(BaseModel):
    id: int
    threat_id: int
    risk_id: int
    # Display names for both ends, embedded by the list/create services so the
    # UI never falls back to raw ids (docs/agent/FRONTEND_DISPLAY_GUARDRAILS.md).
    threat_name: str | None = None
    risk_id_code: str | None = None
    risk_name: str | None = None
    capabilities: ThreatRiskLinkCapabilities | None = None
    created_at: UtcAwareDatetime

    model_config = {"from_attributes": True}

File: frontend/src/types/threat.ts
import type { CollectionGroup } from './collection';

export interface ThreatCapabilities {
    can_read: boolean;
    can_update: boolean;
    can_archive: boolean;
    can_restore: boolean;
}

export interface ThreatListCapabilities {
    can_create: boolean;
    can_export: boolean;
}

export interface ThreatFacetOption {
    value: string;
    label: string;
    count: number;
    disabled: boolean;
    selected: boolean;
}

export type ThreatFacetKey =
    | 'lifecycle'
    | 'category'
    | 'relevant_subject'
    | 'has_linked_risk'
    | 'linked_risk_type';

export type ThreatFacets = Partial<Record<ThreatFacetKey, ThreatFacetOption[]>>;

export interface ThreatLookupOption {
    id: number;
    label: string;
    secondary_label?: string | null;
    disabled: boolean;
    count?: number | null;
}

export interface ThreatSteward {
    name: string;
    email: string;
    role_name: string;
    department_name?: string | null;
}

export type ThreatStewardshipStatus =
    | 'assigned'
    | 'legacy_unassigned'
    | 'pending_governance'
    | 'invalid_assignment';

/** ICT Register Threat — the entered 12_Hrozby columns (issue #47). */
export interface Threat {
    id: number;

    name: string;
    threat_steward_user_id?: number | null;
    threat_steward?: ThreatSteward | null;
    steward_orphaned?: boolean;
    stewardship_status: ThreatStewardshipStatus;
    category?: string | null;
    description?: string | null;
    typical_weaknesses?: string | null;
    relevant_subject?: string | null;
    notes?: string | null;
    is_archived: boolean;
    archived_at?: string | null;
    archived_by_id?: number | null;
    capabilities?: ThreatCapabilities | null;
    created_at: string;
    updated_at: string;
}

export interface ThreatListItem extends Threat {
    visible_linked_risk_count: number;
}

export interface ThreatWritePayload {
    name?: string;
    threat_steward_user_id?: number | null;
    category?: string | null;
    description?: string | null;
    typical_weaknesses?: string | null;
    relevant_subject?: string | null;
    notes?: string | null;
}

export interface ThreatListParams {
    offset: number;
    limit: number;
    search?: string;
    include_archived?: boolean;
    sort_by?: ThreatSortField;
    sort_order?: 'asc' | 'desc';
    sort?: { field: string; direction: 'asc' | 'desc' };
    lifecycle?: Array<'active' | 'archived'>;
    view?: ThreatRegisterView;
    group_by?: ThreatGroupBy;
    group_value?: string;
    categories?: string[];
    steward_ids?: number[];
    relevant_subjects?: string[];
    has_linked_risk?: boolean;
    linked_risk_ids?: number[];
    linked_risk_types?: string[];
    linked_risk_department_ids?: number[];
}

export type ThreatRegisterView = 'all' | 'category' | 'threat_steward' | 'relevant_subject' | 'linked_risk';
export type ThreatGroupBy = Exclude<ThreatRegisterView, 'all'>;
export type ThreatSortField =
    | 'name'
    | 'category'
    | 'threat_steward'
    | 'relevant_subject'
    | 'linked_risk_count'
    | 'created_at';

export interface ThreatListResponse {
    items: ThreatListItem[];
    total: number;
    offset: number;
    limit: number;
    capabilities?: ThreatListCapabilities | null;
    groups?: CollectionGroup[] | null;
    facets?: ThreatFacets | null;
}

export interface ThreatRiskLinkCapabilities {
    /** Mutations follow the MANAGING end: threats:write on the Threat page, risks:write on the Risk detail. */
    can_delete: boolean;
}

/** Threat<->Risk Link relation (issue #47), manageable from both ends. */
export interface ThreatRiskLink {
    id: number;
    threat_id: number;
    risk_id: number;
    /** Server-resolved display names (FRONTEND_DISPLAY_GUARDRAILS: no raw-id fallbacks). */
    threat_name?: string | null;
    risk_id_code?: string | null;
    risk_name?: string | null;
    capabilities?: ThreatRiskLinkCapabilities | null;
    created_at: string;
}

export interface RiskProcessLinkCapabilities {
    /** Mutations follow the Risk end (risks:write). */
    can_delete: boolean;
}

/** Risk<->Process Link relation (issue #47), managed from the Risk detail. */
export interface RiskProcessLink {
    id: number;
    risk_id: number;
    process_id: number;
    /** Server-resolved display names (FRONTEND_DISPLAY_GUARDRAILS: no raw-id fallbacks). */
    process_name?: string | null;
    risk_id_code?: string | null;
    risk_name?: string | null;
    capabilities?: RiskProcessLinkCapabilities | null;
    created_at: string;
}

export interface RiskAssetLinkCapabilities {
    /** Mutations follow the Risk end (risks:write). */
    can_delete: boolean;
}

/** Risk<->Asset Link relation (issue #47), managed from the Risk detail. */
export interface RiskAssetLink {
    id: number;
    risk_id: number;
    asset_id: number;
    /** Server-resolved display names (FRONTEND_DISPLAY_GUARDRAILS: no raw-id fallbacks). */
    asset_name?: string | null;
    risk_id_code?: string | null;
    risk_name?: string | null;
    capabilities?: RiskAssetLinkCapabilities | null;
    created_at: string;
}
